File: I2_CameraMapsApp/app/src/main/java/com/example/tsvetelinastoyanova/cameramapsapp/data/local/SchedulerProvider.java
package com.example.tsvetelinastoyanova.cameramapsapp.data.local;

import io.reactivex.Scheduler;
import io.reactivex.android.schedulers.AndroidSchedulers;
import io.reactivex.schedulers.Schedulers;

public class SchedulerProvider implements Provider {
    @Override
    public Scheduler provideUiThread() {
        return AndroidSchedulers.mainThread();
    }

    @Override
    public Scheduler provideBackgroudThread() {
        return Schedulers.io();
    }
}

File: I3_HackerNewsApp/app/src/main/java/com/example/tsvetelinastoyanova/hackernewsapp/common/schedulers/SchedulerProvider.kt
package com.example.tsvetelinastoyanova.hackernewsapp.common.schedulers

import io.reactivex.Scheduler
import io.reactivex.android.schedulers.AndroidSchedulers
import io.reactivex.schedulers.Schedulers

class SchedulerProvider : BaseSchedulerProvider {

    companion object {
        @Volatile
        private var INSTANCE: SchedulerProvider? = null

        @Synchronized
        fun getInstance(): SchedulerProvider {
            return INSTANCE ?: synchronized(this) {
                INSTANCE ?: SchedulerProvider().also { INSTANCE = it }
            }
        }
    }

    override fun computation(): Scheduler {
        return Schedulers.computation()
    }

    override fun io(): Scheduler {
        return Schedulers.io()
    }

    override fun ui(): Scheduler {
        return AndroidSchedulers.mainThread()
    }
}